### PR TITLE
chore(ruleset): improve then validation

### PR DESCRIPTION
**Pull request by P0lip**
This is a complementary PR to https://github.com/stoplightio/spectral/pull/1136

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

Enhances the ruleset validation to make sure all known `then` with a built-in function are validated according to the typings.
The benefit of that change is the lack of need for any types clobbering as well as any other custom validation (like we had to do in case of casing fn, for instance)

### CODE REVIEW

File: src/meta/rule.schema.json [hunk 1]
```diff
@@ -1,150 +1,464 @@
...
+          },
+          "then": {
+            "properties": {
+              "functionOptions": {
+                "additionalProperties": false,
+                "properties": {
+                  "values": {
+                    "type": "array",
+                    "items": {
+                      "type": ["string", "number"]
```
Comment (nulltoken) [inline src/meta/rule.schema.json:97]: @P0lip Note: Might be worth extending the doco on the restricted types

Comment (P0lip) [reply src/meta/rule.schema.json:97]: Which documentation you mean?
The one about built-in functions?

Comment (nulltoken) [reply src/meta/rule.schema.json:97]: @P0lip Yes. This was what I was thinking about

**Comment by P0lip** [comment]
@nulltoken bump :slightly_smiling_face:

**Comment by P0lip** [comment]
Since there is only the doc change that hasn't been reviewed, I'll go ahead and merge it.
IF you have any concerns about it, LMK and I'll address it in another PR.

**Pull request merged**